seismic/inversion/wavefield_decomp/example_solvers.py

```diff
@@ -0,0 +1,135 @@
+    plt.show()
+# end func
+
+
+def main():
+    # Test functions as per https://en.wikipedia.org/wiki/Test_functions_for_optimization
+    from landscapes.single_objective import sphere, himmelblau, easom, rosenbrock, rastrigin
+
+    logger = logging.getLogger(__name__)
+
+    logging.info("Solving sphere function")
+    bounds = Bounds(np.array([-3, -3]), np.array([3, 3]))
+    soln = optimize_minimize_mhmcmc_cluster(
+        sphere, bounds, burnin=10000, maxiter=50000, collect_samples=10000, logger=logger)
+    if soln.success:
+        logging.info("Solution:\n{}".format(soln.x))
+        plot(soln, title='Sphere function minima')
+    # end if
+
+    logging.info("Solving himmelblau function")
+    bounds = Bounds(np.array([-5, -5]), np.array([5, 5]))
+    soln = optimize_minimize_mhmcmc_cluster(
+        himmelblau, bounds, burnin=10000, maxiter=50000, collect_samples=10000, target_ar=0.3, T=10, logger=logger)
+    if soln.success:
+        logging.info("Solution:\n{}".format(soln.x))
+        plot(soln, title='Himmelblau function minima')
+    # end if
+
+    logging.info("Solving easom function")
+    bounds = Bounds(np.pi + np.array([-3, -2]), np.pi + np.array([3, 4]))
+    soln = optimize_minimize_mhmcmc_cluster(easom, bounds, burnin=10000, maxiter=50000, collect_samples=10000, T=0.2)
+    if soln.success:
+        logging.info("Solution:\n{}".format(soln.x))
+        plot(soln, title='Easom function minima')
+    # end if
+
+    logging.info("Solving rosenbrock function")
+    bounds = Bounds(np.array([-4, -4]), np.array([4, 4]))
+    soln = optimize_minimize_mhmcmc_cluster(rosenbrock, bounds, burnin=10000, maxiter=50000, collect_samples=10000,
+                                            T=1000)
+    if soln.success:
+        logging.info("Solution:\n{}".format(soln.x))
+        plot(soln, title='Rosenbrock function minima')
+    # end if
+
+    logging.info("Solving rastrigin function")
+    bounds = Bounds(np.array([-4, -4]), np.array([4, 4]))
+    soln = optimize_minimize_mhmcmc_cluster(lambda xy: rastrigin(xy) - 2, bounds, burnin=10000, maxiter=50000,
+                                            collect_samples=10000, T=5, N=5, logger=logger)
+    if soln.success:
+        logging.info("Solution:\n{}".format(soln.x))
+        plot(soln, title='Rastrigin function minima')
+    # end if
+
+    # Custom test function
+    logging.info("Solving bi-variate quadratic function")
+    mu = np.array([0, 1])
+    cov = np.array([[5, -6.0], [-6.0, 20.0]])
+    fixed_args = (mu, cov)
+    bounds = Bounds(np.array([-3, -2]), np.array([3, 4]))
+    soln = optimize_minimize_mhmcmc_cluster(bi_quadratic, bounds, fixed_args, burnin=10000, maxiter=50000,
+                                            rnd_seed=20200220, logger=logger)
+    if soln.success:
+        logging.info("Solution:\n{}".format(soln.x))
+        plot(soln, title='Custom bi-variate quadratic function minima')
+    # end if
```

Comment:
Probably we should have some type of tracking of parameters in case there is no success

Reply:
Good suggestion. There is allowance for this in the solution structure returned by the solver, although no detectable failure modes exist at this point. I will add them as they come up.

Reply:
To be clear, this `soln` variable returned by the solver contains a rich amount of information already about the solution process. It is more than just the solution location in parametric space.